### DISCUSSION TITLE
YOLO Performance Metrics - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#6735] [Docs]
# YOLO Performance Metrics - Ultralytics YOLOv8 Docs

A comprehensive guide on various performance metrics related to YOLOv8, their significance, and how to interpret them.

https://docs.ultralytics.com/guides/yolo-performance-metrics/

**Comment by kholidiyah** [comment]
I want to analyze F1-score that get from Yolov8 training, how do i get the value of F1-score and bitrate from training process?

**Reply by pderrenger** [reply]
@kholidiyah during the training process with YOLOv8, the F1-score is automatically calculated and logged for you. To obtain the F1-score and other metrics such as precision, recall, and mAP (mean Average Precision), you can follow these steps:

1. Ensure that you have validation enabled during training by setting `val: True` in your training configuration.
2. After each training epoch, the model will be evaluated on the validation dataset, and performance metrics will be calculated.
3. The results, including the F1-score, will be displayed in the terminal and saved to the `results.csv` file in your run directory (`runs/train/exp*/results.csv` by default).
4. To analyze the F1-score, you can open the `results.csv` file and look for the column labeled `F1` to see the F1-score for each epoch.

As for the bitrate, if you are referring to the model size or the inference speed, these are not directly referred to as "bitrate" in the context of YOLOv8. However, you can find the model size (in megabytes) as a `.pt` file in your run directory, and you can benchmark the inference speed using the `benchmark` mode with the desired settings.

If you need further details on how to interpret these metrics or how to use the `benchmark` mode, please refer to the Ultralytics documentation.

**Comment by mohammadmahdimajidi15-cloud** [comment]
import pandas as pd

df = pd.read_csv("runs/detect/train/results.csv")

last = df.iloc[-1]

final_table = pd.DataFrame({
    "Metric": ["Precision", "Recall", "mAP@0.5", "mAP@0.5:0.95", "F1-Score"],
    "Value": [
        last["metrics/precision(B)"],
        last["metrics/recall(B)"],
        last["metrics/mAP50(B)"],
        last["metrics/mAP50-95(B)"],
        last["metrics/F1(B)"] if "metrics/F1(B)" in df.columns else None
    ]
})

print(final_table)
final_table.to_csv("output/yolo_final_table.csv", index=False)

**Reply by glenn-jocher** [reply]
`results.csv` does not normally include `metrics/F1(B)`, so compute it from the final precision and recall: `p = last["metrics/precision(B)"]; r = last["metrics/recall(B)"]; f1 = 2 * p * r / (p + r) if p + r else 0.0`. Also ensure `output/` exists before calling `to_csv()`. See the [Ultralytics metrics reference](https://docs.ultralytics.com/reference/utils/metrics) for details.
